### ISSUE TITLE
Broker authorization proposal

### ISSUE THREAD

**Issue by kleunen** [redboltz/mqtt_cpp#779]
I had an idea how to add topic authorization to the broker. I would like to propose this idea. 
To start with, you need a database of accounts with possible topic filters as follows:

USER1: Password1
topic: example/+/test, rights: publish

USER2: Password2
topic: example/+/test, rights: subscribe

USER3: Password3
topic: example/+/test, rights: publish + subscribe

So accounts get a list of users with passwords, and topic filters with rights if they are allowed to publish/subscribe to a topic. 

Now, in the broker, when a connect enters:
https://github.com/redboltz/mqtt_cpp/blob/master/include/mqtt/broker/broker.hpp#L405-L450

Rather than calling 'connect_handler' directly:
https://github.com/redboltz/mqtt_cpp/blob/master/include/mqtt/broker/broker.hpp#L439-L448

You pass the connect request to some authorization class:
````C++
class AuthorizerInterface {
  virtual void authorize_connect(
        con_sp_t spep,
        buffer client_id,
        optional<buffer> /*username*/,
        optional<buffer> /*password*/,
        optional<will> will,
        bool clean_start,
        std::uint16_t /*keep_alive*/,
        v5::properties props
    )
}
````

This will lookup the username/password in the databae (possibly a json file or some external authenticator). And finally forward the request to the connect handler within the broker with the rights:

````C++
  bool connect_handler(
        con_sp_t spep,
        buffer client_id,
        optional<buffer> /*username*/,
        optional<buffer> /*password*/,
        optional<will> will,
        bool clean_start,
        std::uint16_t /*keep_alive*/,
        v5::properties props, 
        std::vector< MQTT_NS::buffer > user_publish_filters,
        std::vector< MQTT_NS::buffer > user_subscribe_filters,
    )
````

The user rights are stored in a subscription map, such that we know for each session which rights apply (first set is list of sessions that are allowed to publish, second is list of sessions which are allowed to subscribe):
using sub_rights_map = multiple_subscription_map<buffer, std::pair< std::set<session_state_ref>, std::set<session_state_ref> > >;

Now, when a message is published in 
````C++
bool publish_handler(
        con_sp_t spep,
        optional<packet_id_t> packet_id,
        publish_options pubopts,
        buffer topic_name,
        buffer contents,
        v5::properties props) {
````

Lookup the topic_name in sub_rights_map. 
The publisher should have rights: Publish
The sessions that receive the messages should have rights: Subscribe 

The publisher should be somewhere in any of the filters which is allowed to publish to the topic. 

You can lookup the set of subscribers by looking up the complete set of sessions which are allowed to subscribe to this topic std::set < session_state_ref> >, and then calculating the intersection with sessions which are actually subscribed to this topic.

**Comment by kleunen** [comment]
An alternative would be that on a subscribe you calculate which filters the session is actually allowed to subscribe. So you combined the authorization filter + requested topic filter into an authorized topic filter:

````
Filter:    level1/level2/level3
Subscribe: level1/level2/level3
Result:    level1/level2/level3

Filter:    level1/+/level3
Subscribe: level1/level2/level3
Result:    level1/level2/level3

Filter:    level1/level2/level3
Subscribe: level1/+/level3
Result:    level1/level2/level3

Filter:    level1/+/level3
Subscribe: level1/+/level3
Result:    level1/+/level3

Filter:    level1/level2/#
Subscribe: level1/level2/#
Result:    level1/level2/#

Filter:    level1/level2/#
Subscribe: level1/level2/level3/level4
Result:    level1/level2/level3/level4

Filter:    level1/level2/+
Subscribe: level1/level2/level3/level4
Result:    Not allowed

Filter:    level1/level2/level3/level4
Subscribe: level1/level2/+
Result:    level1/level2/level3/level4

Filter:    level1/+/level3/+
Subscribe: level1/level2/+/level4
Result:    level1/level2/level3/level4
````

I think the rules for this should be:`
````
<literal> + <literal> = <literal>
<literal> + '#' = <literal>
'#' + <literal> = <literal>
<literal> + '+' = <literal> (if number of levels following is the same)
'+' + <literal>= <literal> (if number of levels following is the same)
````

**Comment by kleunen** [comment]
But problem with this approach is, you may get multiple matching filters:

````
    authorized_filters = {
        "example/+/test/+",
        "example/value/+/value"
    };
    
    result = authorize_subscribe(authorized_filters, "example/value/+/value");
    for(auto const& i: result) {
        std::cout << i << std::endl;
    }
````

Results in the following topics to be authorized:
````
example/value/test/value
example/value/+/value
````

In case a publish happens to 'example/value/test/value', this will result in two subscriptions matching. So you have to filter out this duplicate.

**Comment by redboltz** [comment]
mosquitto has its ACL.
http://www.steves-internet-guide.com/topic-restriction-mosquitto-configuration/

If it is good enough, we should use it because it already exists and used. 
User can use existing ACL for mosquitto without conversion. It is a big advantege.

I'm not sure mosquitto ACL is good enough. What do you think?

**Comment by kleunen** [comment]
> mosquitto has its ACL.
> http://www.steves-internet-guide.com/topic-restriction-mosquitto-configuration/
> 
> If it is good enough, we should use it because it already exists and used.
> User can use existing ACL for mosquitto without conversion. It is a big advantege.
> 
> I'm not sure mosquitto ACL is good enough. What do you think?

I did not see your reply. 

Well, the biggest choice to make is: 
- Do you insert subscribes into the subscription map, and when a message is published, check if a subscribed is actually authorized to receive the message

or: 
- When you obtain a subscribe request, do you calculate which topics a user is actually allowed to subscribe ? 

In this case, no unauthorized subscribes will be in the subscription map. 
But, a single subscribe, may result in multiple entries in the subscription map.

**Comment by kleunen** [comment]
Consider the following example: 

User U1 connected to the server and subscribes to topic:
example/#

The ACL for user U1 specified:
example/a (read)
example/b (read)

**Option 1:** When you do ACL check on publish, the subscribe 
U1:example/#
is added to the subscription map. 

And when a message is published to:
example/c

User U1 is subscribed, the ACL for U1 is checked, and U1 will not receive the message

**Option 2:** When you do ACL check on subscribe, the subscribe
U1:example/#

Will be combined with the ACL of user U1 (example/a, example/b)
and the following entries are added to the subscription map:
U1:example/a
U1:example/b

In case example/c is published, user U1 is not matched/subscribed to this topic. 
No ACL check needs to be done on publish

**Comment by kleunen** [comment]
> I'm not sure mosquitto ACL is good enough. What do you think?

What I don't like about mosquittos file format is that you need to write a parser for this format. 

If you use json or ini, you can use boost property tree to parse:
[Boost Property tree](https://www.boost.org/doc/libs/1_75_0/doc/html/property_tree/parsers.html)

For example in mosquitto you can configure:
````
topic read $SYS/#

user roger
topic foo/bar

pattern write $SYS/broker/connection/%c/state
````

In json this would be:
````
{
	"acl": {
		"general": [{
			"topic": "$SYS/#",
			"rights": "read"
		}],
		"users": [{
			"user": "roger",
			"topics": [{
				"topic": "foo/bar",
				"rights": "readwrite"
			}]
		}],
		"patterns": [{
			"pattern": "$SYS/broker/connection/%c/state",
			"rights": "write"
		}]
	}
}
````

**Comment by kleunen** [comment]
Although parsing mosquitto acl fotmat is not that difficult

**Comment by redboltz** [comment]
Sorry, I'm busy on other projects. I will back to mqtt_cpp but I need time.

Just leave some comments for now.

I used to use Boost Property Tree but I don't think it is a good choice.
The design is pretty old-fashioned and not well maintained recently.
If we use json library there are some of better one.

If we need to write parser, Boost.Spirit X3 is good choice for me.
I wrote msgpack parser using it.
See
https://github.com/msgpack/msgpack-c/blob/cpp_master/include/msgpack/v2/x3_parse.hpp
And other files start with `x3_` in the same directory.
I guess that mosquitto ACL is easier than msgpack parsing.
I just want to comment that writing parser is not a problem. It is a natural and straight forward way.

I will read carefully https://github.com/redboltz/mqtt_cpp/issues/779#issuecomment-768944957 later. I don't read it yet.

**Comment by kleunen** [comment]
No problem. I am also quite busy recently. 

Well, the mosquitto ACL format is easy to parse. You can just read lines and use a tokenizer to read them. One thing I don't like is that it does not have clear "sections". 

So you say 'user roger', and all following topics belong to this user:
````
user roger
topic foo/bar

user wouter
topic yadie
````

But if I accidently delete 'user wouter', suddenly 'topic yadie' belongs to 'user roger'. 

Also, what is missing is to limit the pattern of the client id for a user. For example, if user 'roger' logs in, he can only select a client id starting with 'roger_'. This prevents users from having a client_id collision. 

But it is convenient to have the same format as mosquitto, even if you have more options. 

JSON is quite verbose, if property_tree is not an option, There is Boost.JSON, but it is only from version 1.75:
https://www.boost.org/doc/libs/1_75_0/libs/json/doc/html/index.html

And adding another dependency to the project for a JSON parser, i don't think that is a good idea. 

I use property_tree a lot for configuration files, it is convenient. 
Have a config file parser in just a couple of lines. I actually added parsing the uci format to property_tree:
[UCI](https://openwrt.org/docs/guide-user/base-system/uci)

**Comment by kleunen** [comment]
> If we need to write parser, Boost.Spirit X3 is good choice for me.

Ah yes, Spirit X3 uses EBNF, that is quite easy to write. 
Seems quite straightforward, especially for small config format.

**Comment by kleunen** [comment]
Having the configuration in json is convenient, because it is a defacto standard and many tools and libraries exists to create and edit them:
https://linuxhint.com/bash_jq_command/

**Comment by redboltz** [comment]
Sorry to have kept you waiting for a long time.
I had busy days but now, I am back. 

I'd like to confirm the concept of authentication and authorization.
I think that they are separated topic.
I use the word "authentication" as CONNECT permission. The broker somehow checks the connecting client has the rights for connection.
I use the word "authorization" as PUBLISH/SUBSCRIBE/UNSUBSCRIBE permission. The broker somehow checkes the connected client has the rights for PUBLISH and/or [UN]SUBSCRIBE.

There are out of MQTT spec. So we need to decide our broker's spec.

I'd like to start with authentication because it is earlier phase than authorization.

## Authentication
MQTT has Client Identifier(mandatory but empty string has special meaning), Username(optional), and Password(optional).
They are in CONNECT packet.
In addition, since MQTT v5.0, AUTH packet is introduced.
Before MQTT CONNECT packet is sent, we could use client certificate authentication on the TLS layer.
Recently, I added an example for client certificate authentication.
https://github.com/redboltz/mqtt_cpp/blob/d163f887cfa06a950ec7ba3bfd22279680a567d0/example/tls_both_client_cert.cpp#L347-L367

The spec said that Client Identifier must be unique in the broker.
> Each Client connecting to the Server has a unique ClientID.

https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901059

It's very important. If the client that has the same ClientID of the existing connected client, then only one of them is accepted.
The typical implementation is disconnecting the existing client and accepting the new client.
Our broker behaves so.

ClientID could be empty string. In this case, the broker generates unique string and regard it as the ClientID.

On MQTT v5.0, the generated ClientID is notified the client using `3.2.2.3.7 Assigned Client Identifier` property.
https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc511988564

On MQTT v3.1.1, there is no way to notify the generated ClientID to the client.
http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc385349242

Because of this limitation, Clean Session:0 is not allowed with the empty ClientID on MQTT v3.1.1.
> If the Client supplies a zero-byte ClientId, the Client MUST also set CleanSession to 1 [MQTT-3.1.3-7].
> 
> If the Client supplies a zero-byte ClientId with CleanSession set to 0, the Server MUST respond to the CONNECT Packet with a CONNACK return code 0x02 (Identifier rejected) and then close the Network Connection [MQTT-3.1.3-8].
> 
> If the Server rejects the ClientId it MUST respond to the CONNECT Packet with a CONNACK return code 0x02 (Identifier rejected) and then close the Network Connection [MQTT-3.1.3-9].

### How to avoid accidental ClientID conflict
We can choose multiple authentication method candidates.
The first candidate is a combination of `Username` and `Password`.
It's simple but has the following problem.

1. Client1 CONNECT using Username:u1, Password:p1, ClientID:abc
2. Client2 CONNECT using Username:u2, Password:p2, ClientID:abc

The broker disconnects Client1 because Client2 has the ClientID:abc.
Client1 and Client2 could be independent. In this case, they don't know each other.
How to avoid ClientID conflict?

Enforce clients to create UUID and set it as the ClientID for the clients? It is difficlut for embedded (only have poor resornces) devices. That kind devices sometimes don't have system clock. And the system clock is usually used creating random UUID.

Enforce clients to set the only empty ClientID? In this case, the broker can create the unique ClientID. However, on MQTT v3.1.1, clients can use Clean Session:1 only. It is unacceptable. One of big advantage of using MQTT is session continuation.
Note that the session (subscription status, inflight published messages, and offline published messages) are distinguished by ClientID.

How about using ClientID for authentication instead of Username?
1. Client1 CONNECT using ClientID:u1, Password:p1, no Username is used. 
2. Client2 CONNECT using ClientID:u2, Password:p2, no Username is used. 

It works expectedly.

Any ideas?

**Comment by kleunen** [comment]
I think it is best to always user the username for authentication. A username/password combination may be used to connect multiple times. Like you said, a unique ClientID can be connected only once. But it is useful to use the username/password combination to connect multiple devices. 

You can limit the clientid to a certain pattern in many brokers. So if a user logs in with 'username' the valid clients starts with 'username-', so for example: username-0000, username-0001, username-0002 are all valid client ids for this user. This way, no collision between clientids of different users can occur.

**Comment by kleunen** [comment]
For embedded devices i would recommend using the MAC Address or some unique microcontroller identifier. Then the clientid would become: 

username-device.mac

**Comment by redboltz** [comment]
`Username-ClientID` is interesting idea.

Let me clarify what you mean.

1. Client1 CONNECT using Username:u1, Password:p1, ClientID:abc
2. Client2 CONNECT using Username:u2, Password:p2, ClientID:abc

In the broker (mqtt_cpp broker), the internal ClientIDs are `u1-abc` and `u2-abc`. They are different.
So both connections are accepted.
Is that right?

**Comment by kleunen** [comment]
Exactly. 

In mosquitto you have patterns like '%u' which will match the username, so you use them to limit the client id. So one can configure how the clientid handling should look like. I don't think they have a generated pattern, but you might define one '%g'. 

So: 
clientid = %u
clientid = %g
clientid = %u-%g

**Comment by redboltz** [comment]
Ok. I like unconflictable ClientID management.
Let me confirm it.
ClientID is a little bit ambiguous.
One is the Client Identifier field in the CONNECT packet. It is sent from the client. Let's call it scid (sent ClientID).
The other is broker internal identifiler. Let's call it bcid (broker ClientID).
bcid is used for the client conflict checking.

## Case1
The client send CONNECT { Username:"u1", Password:"p1", Client Identifier:"abc" }.
In this case,
scid: "abc"
bcid: "u1-abc"
Only "u1" and "p1" are used for authentication .

## Case2
The client send CONNECT { Username:"u1", Password:"p1", Client Identifier:"" }.
In this case,
scid: ""
bcid: "u1-generated_string_by_broker"
Only "u1" and "p1" are used for authentication .

----
So far, it is good for me.
I have two questions.
1. In MQTT v5.0, what string should be returned as Assigned Client Identifier property in CONNACK packet?
    I think that it should be `generated_string_by_broker`. Because if the client want to inherit the (generated) session on the re-connection, the client would send CONNECT { Username:"u1", Password:"p1", Client Identifier:"generated_string_by_broker" }. The broker would create bcid as `u1-generated_string_by_broker` from the CONNECT packet. It is simple behavior. What do you think?
2. Conformance of MQTT spec.
    The MQTT spec said that "Each Client connecting to the Server has a unique ClientID". The ClientID means scid. In my example case, the same two ClientID "abc" are accepted. I strongly believe that this behavior is better than disconnecting the existing connection that is authenticated by the different Username and Password. So some MQTT Spec lower might say mqtt_cpp broker violates the spec. I'd like to say it is intentional because it is always better behavior. Do you agree ?

**Comment by kleunen** [comment]
I dont think it is a good idea to have an internal and external client id. The client id is also used to continue a persistent session on the broker. So the client may reconnect and identify using the client id and request continuation of a stored session. That is also why the previous possibly already broker connection gets disconnected.

**Comment by redboltz** [comment]
> I dont think it is a good idea to have an internal and external client id. The client id is also used to continue a persistent session on the broker. So the client may reconnect and identify using the client id and request continuation of a stored session. That is also why the previous possibly already broker connection gets disconnected.

I guess that you don't understand what I mean. Because my expression is not well...

> `Username-ClientID` is interesting idea.
> 
> Let me clarify what you mean.
> 
> 1. Client1 CONNECT using Username:u1, Password:p1, ClientID:abc
> 2. Client2 CONNECT using Username:u2, Password:p2, ClientID:abc
> 
> In the broker (mqtt_cpp broker), the internal ClientIDs are `u1-abc` and `u2-abc`. They are different.
> So both connections are accepted.
> Is that right?

You agree to the idea above.
I mean

|client|Username|scid|bcid|
|---|---|---|---|
|client1|u1|abc|u1-abc|
|cleint2|u2|abc|u2-abc|

Do you mean something different?

**Comment by kleunen** [comment]
Ah ok i understand now. Internally you map the client id to a username. So the client wont see this.

Only thing is that when you reconnect with a different username and like to continue a session. This is not possible. But i dont think that is an issue

**Comment by redboltz** [comment]
> Ah ok i understand now. Internally you map the client id to a username. So the client wont see this.
> 
> Only thing is that when you reconnect with a different username and like to continue a session. This is not possible. But i dont think that is an issue

Perfect!
That is I wanted to say!

**Comment by redboltz** [comment]
Let's move on the next topic. It is Client Certificate authentication.
Boost Asio has Client Certificate authentication functionality.
The server(broker) can extract CNAME of the client. (Maybe can also extract Alternative Domain Names)
I think that CNAME (our alt names) should be Username.
The Client Certificate is signed by CA. The broker set CA certificate as verify file.
So if the Client Certificate signature is verified by the broker, then authentication is success.
Important point is Username should be written is the Client Certificate. 

In this case, MQTT's Username and Password shouldn't be used. If the client set them, then disconnect as error or outputs warning logs and ignore them.

ClientID (scid) is used for making multiple sessions using the same authentication.
It is the same as Username, Password, and ClientID(scid) pattern.

It this OK?

**Comment by redboltz** [comment]
Maybe it is effected the ACL.

For example, if the user (client) "redboltz" is used Client Certificate then the ACL becomes something like as follows:

```
{
	"acl": {
		"user": [{
			"roger": {
				"password": "1234"
                         },
                        "redboltz": {
                                "passowrd": "$SYS/ClientCertificate"
                         },
```

```
                                "passowrd": "$SYS/ClientCertificate"
```

means use Client Certificate authentication instead of password authentication.
Or 

```
                                "Client Certificate Auht": true
```

Anyway, it is semantically the same.

**Comment by kleunen** [comment]
> ClientID (scid) is used for making multiple sessions using the same authentication.
> It is the same as Username, Password, and ClientID(scid) pattern.
> 
> It this OK?

Yes. Certificate is alternative for username/password

**Comment by redboltz** [comment]
Ok. 
The rest authentication method is AUTH packet. but It is only for MQTT v5.0. And the spec not said about in detail of AUTH packet. So I think that we can postpone it.

Let's move on authorization step by step.

As we discussed, authentication uses Username and Password, and doesn't use ClientID.

I think that authorization should be defined between topic name (or topic filter) and Username.

If Username u1 has read permission of topic name t1, then any clients that have the username u1 (and may have different ClientIDs) can subscribe t1.
I think that it is reasonable. ClientID could be automatically generated on runtime, it is not suitable for permission control.

What do you think?

**Comment by kleunen** [comment]
> If Username u1 has read permission of topic name t1, then any clients that have the username u1 (and may have different ClientIDs) can subscribe t1.
> I think that it is reasonable. ClientID could be automatically generated on runtime, it is not suitable for permission control.
> 
> What do you think?

yes, agreed. Authentication of topics is based on username or possibly groups of usernames.

**Comment by redboltz** [comment]
Ok, so far, authorization is defined as Username and `topic` relationship.

I've read https://github.com/redboltz/mqtt_cpp/issues/779#issuecomment-768944957 but I'd like to discuss based on my example in order to clarify my understanding.

`topic` is still ambiguous. 
MQTT defines Topic Name and Topic Filter. 
- Topic Filter can contain wildcard charactors '+' and/or '#' but Topic Name can't.
- Topic Name is used on PUBLISH. Topic Filter is used on SUBSCRIBE and UNSUBSCRIBE.
- Topic Name is a kind of Topic Filter but not vice versa.

We can say authorization is Username and Topic **Name** relashonship. How about Topic **Filter**?

Let's say there are following Topic Name:

 ```
example/a
example/b
```

They also are Topic Filter

And there is Topic Filter

```
example/#
```

Authorization table

|Username|Topic (Name or Filter)|Type|Meaning|
|---|---|---|---|
|u1|example/a|read(subscribe)|u1 can subscribe example/a|

In this case, u1 can subscribe `example/a` but cant subscribe `example/b`. So far, very simple.
What happens u1 subscribe `exmaple/#` ?
If we allow only Topic Name in authorization table, then `exmaple/#` should be accepted. And then, move on discussion https://github.com/redboltz/mqtt_cpp/issues/779#issuecomment-768944957 .

But before do that, I'd like to consider the following option.

|Username|Topic (Name or Filter)|Type|Meaning|
|---|---|---|---|
|u1|example/a|read(subscribe)|u1 can subscribe example/a|
|u1|example/#|read(subscribe)|u1 can subscribe example/#|

This is authorization table contains Topic Filter case.
And in this case, u1 can subscribe `example/#` because the (exactly) same topic filter is in the table.
If the table is 
|Username|Topic (Name or Filter)|Type|Meaning|
|---|---|---|---|
|u1|example/a|read(subscribe)|u1 can subscribe example/a|

then u1 can't subscribe `example/#`.

I don't consider deeply yet. 
If you find some problems in the approach, please let me know.

**Comment by kleunen** [comment]
> If the table is
> 
> Username	Topic (Name or Filter)	Type	Meaning
> u1	example/a	read(subscribe)	u1 can subscribe example/a
> then u1 can't subscribe `example/#`.

I think in this case u1 can subscribe to example/#, but will only receive published messages to example/a (and not example/b). 

Because according to mqtt specification, can a subscribe result in a not-authorized response ? If a client subscribes to example/#, is the broker allowed to respond: unauthorized subscription ?

**Comment by kleunen** [comment]
ah yes, it seems it is: 



135 | 0x87 | Not authorized | The Client is not authorized to make this subscription.
-- | -- | -- | --


That does make handling of authorization easier, because you can check on subscription if it is allowed.

**Comment by redboltz** [comment]
> But it can be convenient to subscribe to '#' and only receive the topics you are authorized to receive.

Make sense. 
Before deciding the options, I'd like to consider another issue.

Can authentication table and/or authorization table be updated on runtime?
If we can update them, what happens existing connections and subscriptions?

**Comment by kleunen** [comment]
I would say: yes they can be updated on runtime. Especially if you start adding a "plugin" based authentication system. For example if the authentication rules come from a SQL database. Clearly, they are dynamic. User might like to add or ban users and update authentication for a user, while the broker is running. 

I would say, the updated rules only apply to _new_ sessions. So on startup of a session, the topic authorization is requested for the given username.

**Comment by redboltz** [comment]
Thank you for the comment. I understand we have some choices.
I think that avoiding inconsistent state is important. 
So far, it seems that there is no problem with both applying existing connections (disconnect from the broker, force unsubscribe by the broker, if updated event can be detected) and applying only new connections.

Of course we need some implementation for updating.

**Comment by redboltz** [comment]
Let's back to the authorization issue.

Based on our discussion, I considered the following candidates.

## Candidate1
Authorization table can contain only Topic Name. Topic Filter is not allowed.

```
On SUBSCRIBE received {
   if (Topic Filter contains wildcard) {
       accept the subscription.
   }
   else { // Topic Filter is Topic Name
       lookup authorization table(read)
       if (entry exists) {
           accept the subscription.
       }
       else {
           reject the subscription.
       }
   }
}

```


```
On PUBLISH received {
   lookup authorization table(write)
   if (entry exists) {
       accept the publish. (and continue)
   }
   else {
       reject the publish.
       return
   }
   
   get matched deliver target subscribers (already implemented)
   if (matched subscription is wild card) { // I'm not sure it can be checked or not.
       lookup authorization table(read)
       if (entry exists) {
           deliver PUBLISH packet to the subscriber
       }
       
   }
   else { // no read checking required here
       deliver PUBLISH packet to the subscriber
   }
}
```

## Candidate2
Authorization table can contain only Topic Name and Topic Filter.

```
On SUBSCRIBE received {
   lookup authorization table(read) // wildcard topic filter is also checked
   if (entry exists) {
       accept the subscription.
   }
   else {
       reject the subscription.
   }
}

```

On publish is the same as the Candidate 1


What do you think?

**Comment by kleunen** [comment]
I actually prefer candidate 2 now, because: 

1. The control flow is simpler

2. The client immediatly gets a subscription denied if it is not authorized, instead of silently accepting the subscription and client is not receiving the published messages because it is not authorized

3. On publish you don't have to check subscriber authorization, this is already done on subscription. = easier = better performance.

**Comment by redboltz** [comment]
Let me clarify what does the permission `u1 : example/#` mean?
There are two interpretation.

1. Give the permission subscribing TopicFilter `example/#` (exact match string) to u1. But actual delivery is checked using individual permission like `u1: example/a`. In this case, even if u1 subscribes `example/#`, published message to the topic `example/b` is not delivered to u1.
2. Give the permission subscribing TopicFilter `example/#` (exact match string) to u1. And all matched publish topics are delivered to u1. This doesn't need authorization (read) checking on publish to deliver message.

When I wrote https://github.com/redboltz/mqtt_cpp/issues/779#issuecomment-846384438, I assumed 1.
So I wrote　
> On publish is the same as the Candidate 1

It contains on publish authorization (read) check for delivery.

I noticed that we didn't discuss negative list yet.

If you think 2 is better, how to describe treat negative list?
For example,

u1 : example/# read (allow)
u1 : example/b  read (deny)

I'm not sure this negative entry (deny) is acceptable.
If it is acceptable, the meaning is
- u1 can subscribe `example/#`. The published message `example/b` is not delivered to u1
- u1 can't subscribe `example/b`.

Perhaps negative entry describes like as follows
- Username u1
   - authorization
         - Topic Filter
            - example 
                 - default read deny all
                 - allow read `#`
                 - allow read `a`

This means all topics in `example/` denied, but topic filter `example/#` and `example/a` are allowed. In this case, published message `example/b` doesn't deliver to u1.

**Comment by kleunen** [comment]
Good question, 

maybe have only two options: 

- Deny all, and have individual allows:
default deny all
allow read #
allow read a

- Default allow all, and have individual deny:
default allow all
deny read # 
deny read a

Would that be a good option ?

**Comment by kleunen** [comment]
Indeed this options becomes very ambigious:

u1 : example/# read (allow)
u1 : example/b read (deny)

is example/b allowed or denied ? Is it dependent on order, or does deny override deny, or .. 
User will probably not fully understand or make mistakes

**Comment by redboltz** [comment]
> Good question,
> 
> maybe have only two options:
> 
> * Deny all, and have individual allows:
>   default deny all
>   allow read #
>   allow read a
> * Allow all, and have individual deny:
>   default allow all
>   deny read #
>   deny read a
> 
> Would that be a good option ?

I think that those are good options.

The point is 

1. Define the default entry that is deny or allow.
2. Define individual entries that is opposite of the default entry.

I think that it is reasonable rule.
I said about mosquitto at https://github.com/redboltz/mqtt_cpp/issues/779#issuecomment-755138731 .
> I'm not sure mosquitto ACL is good enough. What do you think ?

As you mentioned at https://github.com/redboltz/mqtt_cpp/issues/779#issuecomment-846404621 , mosquitto rule is ambiguous. To be honest, I don't fully understand it.

I think that it is better to define a clearer rule for mqtt_cpp broker.
The rule might be incompatible to mosquitto in some points but it is acceptable.

We need to define the meaning of wildcard in the entry.

I think that the following rule is good.

1. On subscribe checking
   - Check Topic Filter in SUBSCRIBE packet using table (read) . If the result is `allow` then accept the subscription. If the result is `deny` then deny the subscription. (default should be considered)
       - If the Topic Filter contains wildcard, accepting subscription doesn't mean the all matched topics are delivered on publish.
2. On publish checking
   - Check Topic Name in PUBLISH packet using table (write) . If the result is `allow` then accept the publish. If the result is `deny` then deny the publish. (default should be considered)
   - On delivery phase, wildcard matched topic is checked individually using the table (read). In the checking phase, wildcard entry is ignored.

Let me show an the example.


```
default allow all
deny read a
```

The subscriber can subscribe `#` because `default allow all`. But it is just accepted only subscription.
The publisher publish to the topic `a`.  Publish itself is succeeded because `default allow all.
On delivery phase, the topic `a` is matched to `#`, but the published message is NOT delivered to the subscriber because `deny read a`.

The publisher publish to the topic `b`.  Publish itself is succeeded because `default allow all.
On delivery phase, the topic `b` is matched to `#`, the published message is delivered to the subscriber because `default allow all` and no `deny read b` entry exists.

So, wildcard subscription checked twice. The first checking is on subscribe. The second checking is on publish (delivery phase).

What do you think?

**Comment by kleunen** [comment]
Yes, you are right. 
- Subscription needs to be checked if allowed/denied
- On publish, check needs to be done if a more specific rule allow/denies the message

**Comment by redboltz** [comment]
> Yes, you are right, but I think this is only for default allow.
> 
> * Subscription needs to be checked if allowed/denied
> * On publish, check needs to be done if a more specific rule allow/denies the message
> 
> If the default is deny:
> default deny all
> allow read a
> 
> Then no check at publish is needed ?

I think so if you mean on publish delivery phase checking.

> Subscribe to # is denied, subscribe to a is allowed.

Yes.

> default deny all
> allow read a/#
> 
> Subscribe to # is denied, subscribe to a/# or a/b is allowed.

Yes. 
But a publisher publishes to the topic `a/b` at the delivery phase, the message is NOT delivered.
In order to deliver the message `allow read a/b` is also required.

----
By the way, default is for both read and write.
I'm not sure but if read and write can set individually, it might be useful. 

Something like as follows:

default read deny all
default write allow all
allow read a/b
deny write a/b

Semantic is 

- On `default`
   - decide target rule (read/write/both)
      - If `read` is set, then the rule target is subscribe. If `write` is set then the rule target is publish. If not set anything, rule targets are subscribe and publish.
   - apply deny or allow to the decided target

What do you think ?

**Comment by kleunen** [comment]
> Yes.
> But a publisher publishes to the topic `a/b` at the delivery phase, the message is NOT delivered.
> In order to deliver the message `allow read a/b` is also required.

I understand but i think that this is not very practical. If you have for example a multi tenant broker, a broker used by many companies, you might whitelist a path for a company. 
Say you want to give a company a share of the broker, you might allow them to publish and subscribe under company.com/#. But you do not know which topics they want to create

**Comment by redboltz** [comment]
> > Yes.
> > But a publisher publishes to the topic `a/b` at the delivery phase, the message is NOT delivered.
> > In order to deliver the message `allow read a/b` is also required.
> 
> I understand but i think that this is not very practical. If you have for example a multi tenant broker, a broker used by many companies, you might whitelist a path for a company.
> Say you want to give a company a share of the broker, you might allow them to publish and subscribe under company.com/#. But you do not know which topics they want to create

Good point. 
I understand that your idea defines wildcard in the rule as follows:

```
default deny all
allow read company.com/#
```

On subscribe,
`company.com/#` in the rule table means applying the rule (allow or deny) to any topics filter under `company.com/` including wildcard. So the following subscriptions are accepted.

```
company.com/#
company.com/a
company.com/b
company.com/+/c
```

In addition, there is no check on deliver phase.

Am I understanding correctly ?

**Comment by kleunen** [comment]
Yes exactly

**Comment by kleunen** [comment]
And for default rule, i would say explicit set for read and write. Not for both at same time. But is minor thing

**Comment by kleunen** [comment]
Although

default deny all
allow read company.com/#
allow write company.com/#

Is pretty clear

**Comment by redboltz** [comment]
I'm still not sure a little.
I think that `default` is the same as `#`.

> default deny all
> allow read company.com/#
> allow write company.com/#

So the rule above is the same as

deny read #
deny write #
allow read company.com/#
allow write company.com/#

Is this right?

**Comment by kleunen** [comment]
Yes exactly. But default only set once and to #.

**Comment by redboltz** [comment]
We (at least I) omitted Username in the rule intentionally but I assume that the rule is for one user.

My model is as follows:

- Username: u1
   - default deny all
   - allow read company.com/#
   - allow write company.com/#- 
- Username: u2
   - default allow all

Is this the same as you ?

Or, `default` is more globally applied ?

**Comment by kleunen** [comment]
Per username is flexible, but maybe a bit too flexible? Probably all users will get same default. But can be useful to configure per user?

**Comment by redboltz** [comment]
Your model is as follows ?

- Authentication
   - Username: u1
       - Password p1
   - Username: u2
       - Password p2
- Authorization
  - default deny all
  - Username: u1
     - allow read company.com/#
     - allow write company.com/#- 
  - Username: u2
     - allow write other.company.com/#-

**Comment by kleunen** [comment]
Yes i think that is quite clear

**Comment by redboltz** [comment]
I have other question about `#` in the authorization rule. Now, it has different meaning from MQTT wildcard. I guess that wildcard `+` is not allowed in the authorization rule. I think that it is not so meaningful.
Is that right ?

If it is, `#` could be misleading. I think that `*` is better. It is just a syntax issue.
What do you think ?

**Comment by kleunen** [comment]
No it is possible to check subscription with wildcard against auth rule with wildcard. The subscription has to be less specific or equal specific than the the auth rule. I will explain later, because this needs some examples.

**Comment by kleunen** [comment]
You can check given an auth rule + subscription, if the subscription matches the authentication rules. For each token in the auth + subscription, you can check:

| auth | subscription | authenticated |
| --- | --- | --- |
| literal | literal | yes |
| literal | + or # | no |
| + | literal or + | yes |
| + | # | no |
| # | literal or + or # | yes |

so for example:
auth: example/a, sub: example/a, auth: yes
auth: example/a, sub: example/b, auth: no

auth: example/+/a, sub: example/a/a, auth: yes
auth: example/+/a, sub: example/+/a, auth: yes
auth: example/+/a, sub: example/#, auth: no

auth: example/#, sub: example/a, auth: yes
auth: example/#, sub: example/+, auth: yes
auth: example/#, sub: example/#, auth: yes

**Comment by redboltz** [comment]
Thank you for explaining. I understand. It seems to be good rule.
Let's continue the discussion based on this rule.

I wrote the authorization model https://github.com/redboltz/mqtt_cpp/issues/779#issuecomment-846553347
I noticed that it is **Username based**.
Authorization is a relationship between topics (including Topic Filter) and Username.

I think that users want to make a rule that "any users can publish to the topic a/b", but `Username attacker` should be denied.
What is a good way to explain it ?

**Comment by kleunen** [comment]
default allow all
username frequent_publisher 
  deny write a/b

**Comment by redboltz** [comment]
I think that the following situation seems to be practical and difficult to describe rule.

|topic|user|subscribe|
|---|---|---|
|companyA.com/release/#|non_list_users|deny|
|companyA.com/release/#|u1|allow|
|companyA.com/release/#|u2|allow|
|companyA.com/release/#|u3|allow|
|companyA.com/trial/#|non_list_users|allow|
|companyA.com/trial/#|u7|deny|
|companyA.com/trial/#|u8|deny|

CompanyA provide trial topic and relapse topic. 
Trial topic is for trial. Relapse topic for production.
Trial topic is widely allowed for easy trial but some bad behavior users (e.g. too much publish) are denied (negative listed).
Production topic requires individual permission because of security.

I think that it should be able to be expressed by our rule but I think that it is not possible.
( There are many users.)

**Comment by kleunen** [comment]
Maybe have groups of ours and configure a default for a group of users ? 

CompanyA a different group of users and settings as CompanyB

**Comment by redboltz** [comment]
I got some ideas.

1. Topic based authorization is more natural concept than Username based authentication. Because MQTT topic implies tree structure but Username doesn't. So Topic is good for explain organization structure.
2. As I mentioned `companyA` and `companyB` example, organization could have different default policy. And organization is usually structured. So `companyA/division1` and `companyA/division2` could have different policy. I think that recursive rule should be supported.
3. Authentication and Authorization are separated issue basically. One relation is Authorization is for only authenticated users.

Rule structuring idea:

- Authentication
    - `#` allow  (means global default)
    - `companyA/#` deny (means companyA default)
    - `companyA/division1/#` allow (means companyA/division1 default)
    - `companyA/division1/t1` deny
        - users: u1, u2
    - `companyA/division1/t2` deny
        - users: u2, u3
    - `companyA/division2/#` deny (means companyA/division2 default)
    - `companyA/division2/t1` allow
        - users: u1, u2
    - `companyA/division2/t2` allow
        - users: u2, u3
    - `companyB/#` allow (means companyB default)
    - `companyB/division1/t1` deny (omit `companyB/division1/#` means inherit one layer upper default)
        - users: u2, u3

Result example:
|topic\user|u1|u2|u3|u4|
|---|---|---|---|---|
|companyA/division1/t1|d|d|a|a|
|companyA/division1/t2|a|d|d|a|
|companyA/division1/t3|a|a|a|a|
|companyA/division2/t1|a|a|d|d|
|companyA/division2/t2|d|a|a|d|
|companyA/division2/t3|d|d|d|d|
|companyA/division3/t1|d|d|d|d|
|companyB/division1/t1|a|d|d|a|
|companyB/division1/t2|a|a|a|a|
|companyB/division2/t1|a|a|a|a|
|companyC/division1/t1|a|a|a|a|

`a` means allow, `d` means deny.

Note: u4, t3, division3, and companyC are not in the rule tree. They are written to check default rule.

I use the topic ends with `#` as default rule. It can't have users.
I think that it is simple and flexible.

But users might be expanded as follows if needed:

- Authentication
    - `#` allow  (means global default)
        - users: *
    - `companyA/#` deny (means companyA default)
        - users: *
    - `companyA/division1/#` allow (means companyA/division1 default)
        - users: *
    - `companyA/division1/t1` deny
        - users: u1, u2
    - `companyA/division1/t2` deny
        - users: u2, u3
    - `companyA/division2/#` deny (means companyA/division2 default)
    - `companyA/division2/t1` allow
        - users: u1, u2
    - `companyA/division2/t2` allow
        - users: u2, u3
    - `companyB/#` allow (means companyB default)
        - users: *
    - `companyB/division1/t1` deny (omit `companyB/division1/#` means inherit one layer upper default)
        - users: u2, u3

`users: *` means for all users. If this is omitted, the rule assume `users: *`. 

The rule parsed top to bottom, if conflicted rule entry is appeared, then new one overwrites the old one. Maybe outputting warning message is helpful.


Rule writers can write the rules like a branching tree.

**Comment by kleunen** [comment]
It is possible but it sounds a bit complicated to configure and also to use. With the subscription map you should be able to find all rules which apply when you publish a specific topic:

When you have rules:
companyA/division2/#
companyA/division2/t1

and I publish companyA/division2/t1, the subscription map will match 'companyA/division2/t1' and 'companyA/division2/#'. And you need to know which has priority over which. I would say the more specific rule has priority over the less specific rule.

**Comment by redboltz** [comment]
> When you have rules:
> companyA/division2/#
> companyA/division2/t1
> 
> and I publish companyA/division2/t1, the subscription map will match 'companyA/division2/t1' and 'companyA/division2/#'. And you need to know which has priority over which. I would say the more specific rule has priority over the less specific rule.

Yes, more specific one should has higher priority.

I forgot to add read/writ to my list. I mean the rule is for read.

Just I updated it. The combination of topic and user are not changed.

- Authentication
  - read
    - `#` allow  (means global default)
    - `companyA/#` deny (means companyA default)
    - `companyA/division1/#` allow (means companyA/division1 default)
    - `companyA/division1/t1` deny
        - users: u1, u2
    - `companyA/division1/t2` deny
        - users: u2, u3
    - `companyA/division2/#` deny (means companyA/division2 default)
    - `companyA/division2/t1` allow
        - users: u1, u2
    - `companyA/division2/t2` allow
        - users: u2, u3
    - `companyB/#` allow (means companyB default)
    - `companyB/division1/t1` deny (omit `companyB/division1/#` means inherit one layer upper default)
        - users: u2, u3
  - write
    - `#` allow  (means global default)

> companyA/division2/#
> companyA/division2/t1

Back to your case, `u1` can't subscribe `companyA/division2/#` but can subscribe `companyA/division2/t1`. So only `companyA/division2/t1` is added to the subscription map. When `u9` publishes `companyA/division2/t1`, then `companyA/division2/t1` is matched and the message is delivered to `u1`.

**Comment by redboltz** [comment]
Another case, `u1` can subscribe `companyA/division1/#` but can't subscribe `companyA/division1/t1`. 
Let's say, `u1` has subscribed `companyA/division1/#` now.
What happens when `u9` publishes `companyA/division1/t1`?

We can define two meanings.

1. `read` rule is only for subscription. It is not related to deliver. In other wards, authorization and wildcard matching are independent concept. In this case, published message is delivered to `u1`. No delivery time checking required. 
2. `read` rule is not only for subscription but also for delivery. In this case, published message is NOT delivered to `u1`.

I think that 1 is acceptable.

**Comment by redboltz** [comment]
> Another case, `u1` can subscribe `companyA/division1/#` but can't subscribe `companyA/division1/t1`.
> Let's say, `u1` has subscribed `companyA/division1/#` now.
> What happens when `u9` publishes `companyA/division1/t1`?
> 
> We can define two meanings.
> 
> 1. `read` rule is only for subscription. It is not related to deliver. In other words, authorization and wildcard matching are independent concept. In this case, published message is delivered to `u1`. No delivery time checking required.
> 2. `read` rule is not only for subscription but also for delivery. In this case, published message is NOT delivered to `u1`.
> 
> I think that 2 might difficult to implement or might need high cost checking logic (I'm not sure). In this case 1 is a little bit surprising behavior but acceptable (We need to document **authorization and wildcard matching are independent concept**)
> 
> If 2 can be implemented by practical cost, 2 is better.

I noticed that 1 is bad. Only 2 is acceptable.
If `u1` subscribes `#` then all messages are delivered to `u1`. It is bad.
Sorry for the mix up.

**Comment by kleunen** [comment]
Authentication
   * `#` deny read write
     * trial/# allow read write
     * product/# allow read write
       * users: customer

?

**Comment by redboltz** [comment]
## read write usage of typical IoT application
Hmm. In my IoT service developing experience, a user requires both read and write permission to one topic is rare.

Let's say there are sensor, actuator, and controller.
The sensor reports some status. The controller send a request to the actuator based on the reported sensor's status.
It is one of typical scenario.

Usernames are sensor1, actuator1, and controller1.

There are topics as follows.

```
iot_app/sensors/sensor1_status
iot_app/sensors/actuator1_request
```

`sensor1` can publish `iot_app/sensors/sensor1_status`.
`controller1` can subscribe `iot_app/sensors/sensor1_status`.
`controller1` can publish `iot_app/sensors/actuator1_request`.
`actuator1` can subscribe `iot_app/sensors/actuator1_request`.

## applying my notation (read/write separated)
The minimal authorization rule is as follows:

I wrote **Authentication** in the comment above. It should be **Authorization**.
I sometimes got confused.

----

- Authorization
  - read
     - `#` deny
     - `iot_app/sensors/sensor1_status` allow
       - users: `controller1`
     - `iot_app/sensors/actuator1_request` allow
       - users: `actuator1`
  - write
     - `#` deny
     - `iot_app/sensors/sensor1_status` allow
       - users: `sensor1`
     - `iot_app/sensors/actuator1_request` allow
       - users: `controller1`

Note

     - `#` deny

is the short form of

     - `#` deny
         - users: *


There is no read/write allowed user in the one topic.

## applying refined notation based on your comment (read/write mixed)

However, https://github.com/redboltz/mqtt_cpp/issues/779#issuecomment-850870655 's advantage is compact notation. 

----

- Authorization
   - `#` deny
   - `iot_app/sensors/sensor1_status` allow
     - read  `controller1`
     - write `sensor1`
   - `iot_app/sensors/actuator1_request` allow
     - read `actuator1`
     - write `controller1`

Note

     - `#` deny

is the short form of

     - `#` deny
         - read *
         - write *

If `read` and/or `write` is omitted, then regard it as `*` (all users).
For example,

- Authorization
   - `#` deny // deny all users read and write
     - read *
   - `iot_app/sensors/sensor1_status` allow // allow read `controller1` and `controller2`, allow write all users.
     - read  `controller1`, `controller2`
   - `iot_app/sensors/actuator1_request` allow // allow all users read and write
----

What do you think of the last version (refined notation based on your comment (read/write mixed)) ?

**Comment by kleunen** [comment]
Yes, this looks good. But I would make the default:

If read and/or write is omitted, then regard it as nobody (no users).

So make it explicit you want to allow everybody, from security perspective.

**Comment by kleunen** [comment]
But it would be useful to combine users into groups

so you can say: 

group controller_group controller1 controller1
allow controller_group

**Comment by redboltz** [comment]
> Yes, this looks good.
> 
> But I would make the default:
> _If read and/or write is omitted, then regard it as nobody (no users)._
> So make it explicit you want to allow everybody, from security perspective.

Ok, I agree. Let me clarify.
Let's say `*` is all users and `-` is no users.

## Allow omitting

### Case 1
- Authorization
  - `#` deny

is the same as

- Authorization
  - `#` deny
    - read *
    - write * 

Because deny with omitting read/write should be for all users.

### Case 2
- Authorization
  - `#` allow

is the same as

- Authorization
  - `#` allow
    - read `-`
    - write `-` 

Because allow with omitting read/write should be for no users.

Observation.
Default meaning is inconsistent.

## Disallow omitting
- Authorization
  - `#` deny
    - read *
    - write * 

read and write are mandatory. If omitted, then format error is reported.


---
I think that disallow omitting approach is simpler. What do you think?

**Comment by kleunen** [comment]
Maybe not simpler, because user always has to add read and write. But at least it is clear who is allowed. Maybe give in error message if read and write is missing explanation of how to allow all, allow nobody or allow username.

**Comment by redboltz** [comment]
> Maybe not simpler, because user always has to add read and write. But at least it is clear who is allowed. Maybe give in error message if read and write is missing explanation of how to allow all, allow nobody or allow username.

What do you think about Allow omitting rule? deny with omitting means for all users. allow with omitting means for no users. Do you agree this rule ?

**Comment by redboltz** [comment]
> And it would be useful to combine users into groups
> 
> so you can say:
> group controller_group controller1 controller1
> allow controller_group

I agree to introducing user group concept.

One user can be a member of multiple groups.
I considered the following cases.

- user_group
  - group1
    - users: u1, u2
  - group2
    - users: u2, u3
- Authorization
  - `#` deny
    - read *
    - write *
  - `topic1` allow
    - read group1
    - write group1
  - `trial/#` allow
    - read *
    - write *
  - `trial/topic2` deny
    - read group1
    - write group1
  - `messy/#` allow
    - read group1
    - write group2
  - `messy/topic3` deny
    - read group2
    - write group1

And the permission should be as follows:

u1 can read/write `topic1`.
u2 can read/write `topic1`.
u3 can't read/write `topic1`.

u1 can't read/write `trial/topic2`.
u2 can't read/write `trial/topic2`.
u3 can read/write `trial/topic2`.

u1 can read `messy/topic3`.
u2 can't read `messy/topic3`.
u3 can't read `messy/topic3`.
u1 can't write `messy/topic3`.
u2 can't write `messy/topic3`.
u3 can write `messy/topic3`.

Do you think so?

**Comment by kleunen** [comment]
Yes i agree with allow ommiting rule. If default is nobody for allow and everybody for deny. This is ok too. It is also a good approach.

You can block an offending topic easily. Just say deny and nobody can read and write. Can be convenient.

**Comment by kleunen** [comment]
Maybe have ommiting for deny: everbody is denied.

But give error if ommited on allow: please specify who is allowed. All or specific user or group of users.

**Comment by kleunen** [comment]
> > And it would be useful to combine users into groups
> > so you can say:
> > group controller_group controller1 controller1
> > allow controller_group
> 
> I agree to introducing user group concept.
> 
> One user can be a member of multiple groups.
> I considered the following cases.
> 
> * user_group
>   
>   * group1
>     
>     * users: u1, u2
>   * group2
>     
>     * users: u2, u3
> * Authorization
>   
>   * `#` deny
>     
>     * read *
>     * write *
>   * `topic1` allow
>     
>     * read group1
>     * write group1
>   * `trial/#` allow
>     
>     * read *
>     * write *
>   * `trial/topic2` deny
>     
>     * read group1
>     * write group1
>   * `messy/#` allow
>     
>     * read group1
>     * write group2
>   * `messy/topic3` deny
>     
>     * read group2
>     * write group1
> 
> And the permission should be as follows:
> 
> u1 can read/write `topic1`.
> u2 can read/write `topic1`.
> u3 can't read/write `topic1`.
> 
> u1 can't read/write `trial/topic2`.
> u2 can't read/write `trial/topic2`.
> u3 can read/write `trial/topic2`.
> 
> u1 can read `messy/topic3`.
> u2 can't read `messy/topic3`.
> u3 can't read `messy/topic3`.
> u1 can't write `messy/topic3`.
> u2 can't write `messy/topic3`.
> u3 can write `messy/topic3`.
> 
> Do you think so?

Yes. But maybe group name always prefixed with @?

@group1 @group2 u1 u2

**Comment by kleunen** [comment]
But if group is allowed and user is denied? User rule gets priority over group rule?

**Comment by redboltz** [comment]
> Yes. But maybe group name always prefixed with @?
> @Group1 @Group2 u1 u2

Prefix is a good idea. MQTT spec allows any UTF-8 string for Username.
https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901071

So Username might starts with `@`.
Maybe a requirement `mqtt_cpp broker client must not use a Username starts with `@`.

**Comment by redboltz** [comment]
> But if group is allowed and user is denied? User rule gets priority over group rule?

I think that it is a similar situation, `u2` is member of `group1` and `group2`. And `group1` allowed and `group2` denied.
But it couldn't happen at the same topic.

- Ahthorization
  - `#` deny
    - `sub1/#` allow // read group1 write no users(because of omittnig)
       - read @group1
    - `sub1/topic1` deny
       - read u1

In this case u1 can subscribe `sub1/#`, `sub1/any_topics_except_topic1`, but can't subscribe `sub1/topic1`. 
When u1 has subscribed `sub1/#`, then the publish `sub1/topic1` happen, the message is NOT delivered to u1 as we discussed.
There is no additional rule and confliction that is introduced by group concept.

Another case:

- Ahthorization
  - `#` deny
    - `topic1` deny
       - read u1
    - `topic1` allow
       - read @group1

I think that it should be format error. The same topic entry should appear once. 
If it appears twice or more, it should be error.

**Comment by kleunen** [comment]
>   * `#` deny
>     
>     * `topic1` deny
>       
>       * read @Group1
>     * `topic1` allow
>       
>       * read @Group2
> 

And user u1 is part of Group1 and Group2 ? 

I think user should only be part of 1 group.

**Comment by redboltz** [comment]
> * `#` deny
>   
>   * `topic1` deny
>     
>     * read @Group1
>   * `topic1` allow
>     
>     * read @Group2
> 
> And user u1 is part of Group1 and Group2 ?
> 
> I think user should only be part of 1 group.

This is format error.

I assume that the parsing  process is top to bottom.

- `#` deny
   - `topic1` deny // maybe output warning. explicit deny in deny all users is redundant
       - read @Group1 
    - `topic1` allow // error `topic1` appears twice. Should appear only once.
       - read @Group2

I added comments.

**Comment by kleunen** [comment]
What about:

**Comment by redboltz** [comment]
> What about:
> 
> * `#` deny
>   
>   * `/sub/topic1` deny
>     
>     * read @Group1
>   * `/sub/#` allow
>     
>     * read @Group2
> 
> ?

It is the same as follows:

- `#` deny
  - read `*`
  - write `*`
- `/sub/topic1` deny
  - read @Group1
- `/sub/#` allow
  - read @Group2



- `#` deny
  - read `*`
  - write `*`
- `/sub/topic1` deny // redundant because `#` is denied. Let's say rule1
  - read @Group1
- `/sub/#` allow        // Two possible options see below. Lets's say rule2
  - read @Group2

Two possible options.
1. Simply top to bottom parse. So rule2 overwrite rule1. 
2. Sort by wide to narrow. And then apply rules.

After sorted:

- `#` deny
  - read `*`
  - write `*`
- `/sub/#` allow
  - read @Group2
- `/sub/topic1` deny
  - read @Group1

|username|subscribe sub/#|subscribe sub/topic1|deliver (sub/topic1)|
|---|---|---|---|
|u1|no|no|no|
|u2|yes|no|no|
|u3|no|yes|yes|

**Comment by redboltz** [comment]
> more specific topic should follow broader topics ?
> so should be ?

"more specific" is the same meaning as "sort wide to narrow" I commented.

I recommend writing the rule this order. But I think that it can be sorted by the broker. If it can, sort is kinder implementation. And possibly, output warning message if users rule is not sorted.

**Comment by kleunen** [comment]
Yes a warning should be generated if rules are applied in different order

**Comment by redboltz** [comment]
Maybe I edited the comment https://github.com/redboltz/mqtt_cpp/issues/779#issuecomment-850982726 after you read.
Please check it again :)

**Comment by kleunen** [comment]
I think it is ok like this.

**Comment by redboltz** [comment]
Thank you!
I think that the semantics are fixed.

The next step is syntax.
I wrote JSON example:


```js
{
  "authentication": {
    {
      "name": "u1",
      "method": "password",
      "password": "mypassword"
    },
    {
      "name": "u2",
      "method": "client_cert"
    }
  },
  "group": {
    "@g1" : ["u1", "u2"]
  },
  "authorization": [
    {
      "topic": "#",
      "type": "deny"
#     "pub": ["*"]     # can omit
#     "sub": ["*"]     # can omit
    },
    {
      "topic": "sub/#",
      "type": "allow",
      "sub": ["@g1"]
    },
    {
      "topic": "sub/topic1",
      "type": "deny",
      "sub": ["u1"]
    },
  ]
}
```

What do you think?

**Comment by kleunen** [comment]
I think text based rules with tabs is better.

But you can support both

topic /topic1 allow
    read: u1

**Comment by redboltz** [comment]
> I think text based rules with tabs is better.

JSON is a little bit redundant. So simpler notation is nice.
By the way, "tabs" means indent ? I personally don't like TAB character. Indent is good,

If we support JSON and ini file format, you can use boost property tree.
If you want to support other (original) text format, then you need to use Boost.Spirit (or X3).
X3 is more sophisticated but experimental (it actually works).

**Comment by redboltz** [comment]
Or do you know any good library to parse indented text ?

**Comment by kleunen** [comment]
Yes indent is spaces or tabs. 

I do not know any parser. Maybe there if a yaml parser based on spirit?
https://en.m.wikipedia.org/wiki/YAML

**Comment by redboltz** [comment]
Ok, I think that Boost.Spirit.X3 is good one to write parser.

## Document
https://www.boost.org/doc/libs/1_76_0/libs/spirit/doc/x3/html/index.html

## MessagePack format parser (example)
https://github.com/msgpack/msgpack/blob/master/spec.md
https://github.com/msgpack/msgpack-c/blob/cpp_master/include/msgpack/v2/x3_parse.hpp

I think that writing PoC code to parse indented text. It outputs C++ data structure.

**Comment by kleunen** [comment]
You can also parse indented text to property_tree, that way you only have to handle property tree when reading configuration. And also be able to input ini and json format.

**Comment by redboltz** [comment]
Which do you mean Pattern A or Pattern B ?

If you mean B, I think that C is better.
If you mean A, we need to check property_tree has enough accessing method.
`mqtt_cpp_some_data_structure` might be multi_index. It can provide flexible access.
I guess that the data structure needs to have flexible accessing methods if we implement on runtime update in the future.

## Pattern A
```
ini -----------------------+
                           |
                           V
json-----------------> property_tree ---> broker
                           A
                           |
indented text -------> spirit x3
```

## Pattern B
```
ini -----------------------+
                           |
                           V
json-----------------> property_tree ---> mqtt_cpp_some_data_structure ---> broker
                           A
                           |
indented text -------> spirit x3
```

## Pattern C
```
ini -----------------------+
                           |
                           V
json-----------------> property_tree ---> mqtt_cpp_some_data_structure ---> broker
                                                    A
                                                    |
indented text -------> spirit x3 -------------------+
```

**Comment by kleunen** [comment]
Pattern B. 

mqtt_cpp_some_data_structure will be a subscription_map probably, and some combination of datastructures, possibly. You do want to optimize the checking of rules when user logs in. 

std::map<username, userinfo>
std::map<groupname, std::set<username> >
subscription_map<rules, acces rights> 

something like this.

**Comment by redboltz** [comment]
I think that spirit x3's semantic action adds parse result to property_tree (Pattern B) or mqtt_cpp_some_data_structure (Pattern C) repeatedly.
I think that it Pattern C is simpler and straight forward approach.
I'm not sure but I guess that property_tree is designed for parser and element accessor. In pattern B, property_tree is used as container. Maybe insert some element to the property_tree in the semantic action. It is a little weird for me.

**Comment by kleunen** [comment]
Just have a look what is easiest. Maybe first define the internal datastructures for fast authentication and rule matching.

**Comment by redboltz** [comment]
Ok.
By the way, I guess that sorting by wide to narrow will be implemented in mqtt_cpp_some_data_structure. At least property_tree doesn't have such functionality. 

> Maybe first define the internal datastructures for fast authentication and rule matching.

Yes, I think that it is a good way.

**Comment by kleunen** [comment]
Have you made any progress on the authorization ? or not working on mqtt_cpp ?

**Comment by redboltz** [comment]
I'm working on my company's broker and SDKs. So unfortunately, I don't have much time for mqtt_cpp.
I think that the spec of authentication and authorization is almost fixed. At least we have the agreement for the controversy part.
So I think that you can start implementing them. PR is welcome :)

**Comment by kleunen** [comment]
I was just a bit worried, you haven't committed anything since may. I was thinking: i hope nothing bad happened to you. But luckily you are just busy. 

Yes, maybe after my holiday, I may pickup some work again on mqtt_cpp. I have been runner the broker for quite a while now, it is completely stable. Although only used sometimes.

**Comment by redboltz** [comment]
Sorry for making you worry. 

This is one of my recent activity on github. The logic is from mqtt_cpp. (The PR itself created some time ago.)
https://github.com/mqttjs/MQTT.js/pull/1243#issuecomment-865393719
This is a related work of my company's (extended) MQTT SDKs.

**Comment by ineffective** [comment]
First, sorry for not reading everything in this thread. Second - I read your ruminations about reg. certificates. Could you consider adding some way to extract different field than CNAME? We use `UID` (`oid`: `0.9.2342.19200300.100.1.1`) to store client `UUID` and this is what we use for authentication - we currently use `ActiveMQ` with custom plugin, but we consider switching to custom broker (mqtt_cpp based of course) and if this option was there it would be really great.

**Comment by kleunen** [comment]
I guess that would be good to take into consideration when designing an authentication system. If i understand correctly you have a custom authentication plugin for activemq?

**Comment by ineffective** [comment]
Yes, we have custom plugin for activemq. For mqtt connections it simply takes remote certificate used for authentication, extracts `UID` and checks if `client_id` is set to the same value. If yes then uses data from certificate for authorization, otherwise connection is dropped. It would be nice if same thing could be done using `mqtt_cpp` broker. `UID` extraction by `oid` is done using this (maybe a little clumsy, but at least well commented IMHO) code:

```cpp
std::vector<std::byte> get_tag_from_cert(std::string const& file, std::string const& tag_string) {
    // open certificate file
    FILE* fl = fopen(file.c_str(), "rb");
    if (fl != nullptr)
    {
        // if file was opened succesfully, create shared pointer that will automatically close it after it goes out of scope
        boost::shared_ptr<FILE> f(fl, ::fclose);
        // generate OID (object id) for requted tag (parse tag_string)
        ASN1_OBJECT* obj = OBJ_txt2obj(tag_string.c_str(), 1);
        // create certificate object from data read from file
        boost::shared_ptr<X509> x509(PEM_read_X509(f.get(), NULL, NULL, NULL), X509_free);
        // get subject name from the certificate. no need for separate deletion, this is only a pointer to value stored in x509 object
        auto sn = X509_get_subject_name(x509.get());
        // array for UUID read from subject that we got above
        char arr[128] = { 0 };
        // read text value of requested object (obj) in subject name (sn), store in arr (text buffer)
        auto rv = X509_NAME_get_text_by_OBJ(sn, obj, arr, sizeof(arr));
        // check for any errors
        if (rv < 0 || (rv > 0 && static_cast<size_t>(rv) >= sizeof(arr))) {
            // ... and throw if there are any
            throw std::runtime_error("can't get tag [" + tag_string + "] from file [" + file + "], rv: " + std::to_string(rv));
        } else {
            // prepare return value, namely create preallocated vector
            std::vector<std::byte> ret_val{size_t(rv)};
            // copy bytes from arr buffer to return vector (ret_val)
            std::transform(arr, arr + size_t(rv), ret_val.begin(), [](char c) { return std::byte(c); });
            // terminate with NUL byte
            ret_val.push_back(std::byte(0));
            // and return
            return ret_val;
        }
    } else {
        // ... or throw an exception if file couldn't be opened
        throw std::runtime_error("can't open cert file: " + file);
    }
    // automatic storage reclamation will take care of closing file and freeing memory used by the certificate
}
```

**Comment by redboltz** [comment]
CNAME is commonly used. I don't know what UID mean. Is it common field of client certificate?

client_id is different concept, please read all thread.
CNAME (or UID) is for Username. And TLS authentication is same effect as password.
So we don't compare client_id. It is for multiple sessions.

**Comment by ineffective** [comment]
> CNAME is commonly used. I don't know what UID mean. Is it common field of client certificate?

No, it is not common, but it is defined (has `OID` assigned) and is part of `subject name`. You can see it for example here: http://oid-info.com/get/0.9.2342.19200300.100.1.1 and it is also mentioned here: https://www.cryptosys.net/pki/manpki/pki_distnames.html

> client_id is different concept, please read all thread.

I'm trying, but it is *very* long. As far as I understand `client_id` can be set to basically any value by the connecting side - this is what we want to prevent by checking if this value is equal to `UID`, because otherwise there is a collision. Note that this is custom plugin for `ActiveMQ`, not something we get by default. Problem that it is trying to solve is similar (if I understand it correctly) to what @kleunen wrote in one of previous posts:

>> Also, what is missing is to limit the pattern of the client id for a user. For example, if user 'roger' logs in, he can only select a client id starting with 'roger_'. This prevents users from having a client_id collision.

In fact I this is *exactly* what we avoid using this plugin.

> CNAME (or UID) is for Username. And TLS authentication is same effect as password.

`cname` is `common name`. This is logically different from `uid` (so `user id`). In our use-case `cname` contains completely different set of data (parse-able string that contains information like hostname, device type and other - it could be done differently, but it was done this way a long time ago).

> So we don't compare client_id. It is for multiple sessions.
> Authentication method 	Username 	Unique Session Identifier
> MQTT Username Password 	Username 	Username+ClientId(Username is upper)
> TLS Client cert 	CNAME(or other signed field) as Username 	Username+ClientId(Username is upper)

I understand that in our case "other signed field" can be `uid` that I mentioned before. This is exactly what we want. The question is whether this field can be specified using configuration file. If yes, then quite probably `NID` is not the best choice. As far as I understand `NID` is `numeric identifier` which is assigned to specific `OID` and it is `openssl` specific. I'm not sure, but I believe there are no guarantees that `NID` won't change between library releases.

I'm slowly reading this thread and try to decide if this is something we will be able to use out of the box. Quite probably not and quite a lot of programming will be required anyway, so maybe I'm just wasting your precious time. Sorry.

**Comment by redboltz** [comment]
> > > Also, what is missing is to limit the pattern of the client id for a user. For example, if user 'roger' logs in, he can only select a client id starting with 'roger_'. This prevents users from having a client_id collision.
> 
> In fact I this is _exactly_ what we avoid using this plugin.

So we use `Username` * `ClientID` approach. Strictly speaking, it could be violate MQTT specification. But we do it because it is practical approach.

> > CNAME (or UID) is for Username. And TLS authentication is same effect as password.
> 
> `cname` is `common name`. This is logically different from `uid` (so `user id`). In our use-case `cname` contains completely different set of data (parse-able string that contains information like hostname, device type and other - it could be done differently, but it was done this way a long time ago).
> 
> > So we don't compare client_id. It is for multiple sessions.
> > Authentication method 	Username 	Unique Session Identifier
> > MQTT Username Password 	Username 	Username+ClientId(Username is upper)
> > TLS Client cert 	CNAME(or other signed field) as Username 	Username+ClientId(Username is upper)
> 
> I understand that in our case "other signed field" can be `uid` that I mentioned before. This is exactly what we want. The question is whether this field can be specified using configuration file. If yes, then quite probably `NID` is not the best choice. As far as I understand `NID` is `numeric identifier` which is assigned to specific `OID` and it is `openssl` specific. I'm not sure, but I believe there are no guarantees that `NID` won't change between library releases.
> 
> I'm slowly reading this thread and try to decide if this is something we will be able to use out of the box. Quite probably not and quite a lot of programming will be required anyway, so maybe I'm just wasting your precious time. Sorry.

I personally think that Username Password authentication is a good start point, then authorization, and finally, Client Certificate authentication. For now, mqtt_cpp broker doesn't have any authentication authorization ocde.
I don't know much about OpenSSL API (it is very difficult to use for me) . I use NID based field search but replace it with object? based approach is not an essential problem. We can implement it using flexible approach.
In order to switch authentication method, the broker can provide customization point. This requires recompile. Maybe some default implemented authentication selector (that including cert field selector) using boost program options is better.

**Comment by kleunen** [comment]
The thread is very long, many different topics are involved and many possibilities where discussed. To go forward, i think some of the different topics need to be combined into different proposals. These can be discussed separately. I was involved in this discussion, but due to available time, i lost track a bit as well. Possibly i will have some time soon to work on the broker again. 

I do have experience with openssl/ssl based authentication.

**Comment by kleunen** [comment]
> I'm slowly reading this thread and try to decide if this is something we will be able to use out of the box. Quite probably not and quite a lot of programming will be required anyway, so maybe I'm just wasting your precious time. Sorry.

At the moment nothing of authentication and access control is implemented. People work on this project in spare time. Authentication/access control is a big topic and PR accepted in this project only get accepted if quality is very high + tested. So be aware most likely implementing this feature will take significant time. Project focus is also on building a client, not really a broker.

**Comment by redboltz** [comment]
Actually the thread is very long.
Let me show you guideline.

## Authentication
Authentication is permission for connection. It is checked on CONNECT.

Authentication method|Username|Unique Session Identifier
----|----|----
MQTT Username Password|Username|Username+ClientId(Username is upper)
TLS Client cert|CNAME(or other signed field) as Username|Username+ClientId(Username is upper)

### Important note
MQTT spec requires Client Identifier(ClientId) must be unique on one broker.
mqtt_cpp broker intentionally violate it.
On mqtt_cpp Username + ClientId must be unique. Separator is not decided yet.
For example, let's say separator is `-` .

The following two connection can exist on the same time:

Username: User1, ClientId: Id1
Username: User2, ClientId: Id1

It is treat as `User1-Id1` and `User2-Id1` internally.

The following two connections share the same authentication. 
That means the same Username multiple connections can exist if ClientId is different.
Username: User1, ClientId: Id1
Username: User1, ClientId: Id2

How to treat the empty ClientId is not completely decided yet.
https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901059

## Authorization
Authorization is a permission between Username and Topic Filter.
That means `User1-Id1` and `User1-Id2` have the same permission.
It is checked on PUBLISH, SUBSCRIBE, and **deliver**.

On **deliver** checking is important to treat well Wildcard Topic Filter.

Read 10 comments from
https://github.com/redboltz/mqtt_cpp/issues/779#issuecomment-850970590
to
https://github.com/redboltz/mqtt_cpp/issues/779#issuecomment-850982048

**Comment by kleunen** [comment]
Following example loads the json config using property_tree:

https://wandbox.org/permlink/iJmuEhGC3txDK02v

**Comment by redboltz** [comment]
Wandbox supports multiple files. The left most tab is file for `int main() {}`. You can add any files the right of the first file.
https://wandbox.org/permlink/smj7eCPjxcYJhVBN
This demonstrates `data.json`.

Could you update your wandbox code?

**Comment by kleunen** [comment]
https://wandbox.org/permlink/g5k1OF3e8YZ33sPS

**Comment by redboltz** [comment]
> The group format was changed, because with the old format, you can define multiple groups within the same object:
> 
> ```
> 	"group": [{
> 		"name": "@g1",
> 		"members": ["u1", "u2"]
> 	}],
> ```

What does it mean? I don't understand the difference.

The new format can define multiple groups as follows:

```json
	"group": [
        {
	    	"name": "@g1",
		    "members": ["u1", "u2"]
	    },
        {
	    	"name": "@g2",
		    "members": ["u2", "u3"]
	    }
    ],
```

I'm not sure what it the old format. Could you show me the new and old example?

**Comment by kleunen** [comment]
New format: 
````
    "group": [
        {
            "name": "@g1",
            "members": ["u1", "u2"]
        },
        {
            "name": "@g2",
            "members": ["u2", "u3"]
        }
    ],
````

Old format: 
````
    "group": [
        {
            "@g1": ["u1", "u2"],        
            "@g2": ["u2", "u3"]
        },
        {
            "@g3": ["u2", "u3"]
        }
    ],

**Comment by redboltz** [comment]
> I think the password should not be stored plaintext, it should be salt+hash. So one should also be able to configure the salt.

Good point.
In my broker, I use digest, salt, algorithm (for digest), parameter (for algorithm, optional).

**Comment by redboltz** [comment]
I just glanced the code. It seems that the structures (Authentication etc) are directly reflected the json format. It is to store the json. 

What is `BrokerSecurity`'s responsibility ?
I guess that it would be a member of broker.
Let's say broker has a menber variable `BrokerSecurity bs;`.
Maybe something like as follows ?
- `bs.check_user(username, params)` for connect.
- `bs.can_publish(username, topic_filter)` for publish.
- `bs.can_subscribe(username, topic_name)` for subscribe and delivery.

I think that those member functions would be implemented using `authentication`, `authorization`, and `groups`.

BTW, I think that the code is PoC. So it is not satisfied https://github.com/redboltz/mqtt_cpp/wiki/Coding-Rules . It's no problem, so far. In the actual code (PR), please use the Coding-Rules.

**Comment by kleunen** [comment]
Yes. It is just to try out and think about some things. Like do you want to make hash configurable or just sha256 for password. And you want to allow anonymous login? And how? anonymous is a special user?

**Comment by redboltz** [comment]
> Yes. It is just to try out and think about some things. Like do you want to make hash configurable or just sha256 for password. And you want to allow anonymous login? And how? anonymous is a special user?

Configurable hash is good. So far, sha-256 is good enough. But I want to have a customization point in the json format.

For anonymous login, how about this? (Just idea)

## no password login

```js
    "authentication": [
        {
            // no username and no password accepted
        },
        {
            // empty string username with no password accepted
            "name": ""
        },
        {
            // no password accepted
            "name": "u3" 
        },
        {
            // no username with password accepted
            "method": "password",
            "password": "mypassword"
        },
        {
            // empty string username with password accepted
            "name": "",
            "method": "password",
            "password": "mypassword"
        }
    ]
````

**Comment by kleunen** [comment]
Making hash configurable is good I would say, maybe default to sha256 indeed. 

I was thinking something like: 
````
  "authentication": [
        {
            "name": "anonymous",
            "method": "anonymous"
        }
    ]
````

You define for a user the method is anonymous, basicly specifying what the anonymous username should be.

**Comment by kleunen** [comment]
> What is `BrokerSecurity`'s responsibility ? I guess that it would be a member of broker. Let's say broker has a menber variable `BrokerSecurity bs;`. Maybe something like as follows ?
> 
> * `bs.check_user(username, params)` for connect.
> * `bs.can_publish(username, topic_filter)` for publish.
> * `bs.can_subscribe(username, topic_name)` for subscribe and delivery.
> 
> I think that those member functions would be implemented using `authentication`, `authorization`, and `groups`.

Yes, something like this. But the authorization filters should be stored in a subscription_map, for fast lookup.

**Comment by redboltz** [comment]
Sorry for my late replay. I need to implement automatic client_id generation and request response topic.
It could be related to the issue.

When the client CONNECT using empty client_id, then the broker generates unique client_id (using uuid, so far).
When the client CONNECT with RequestResponseInformation(1) properry, then the broke generates unique TopicName (using uuid, so far) and send CONNACK with ResponseTopic property.

Maybe auto generated client_id is not related to authentication and authorization because authentication is based on UserName not clinet_id.
Auto generated TopicName for ResponseTopic is related to authorization.
I think that the ResponseTopics are tagged something like ResponseTopicTag.
ResponseTopic should only be able to subscribe by the client that sends RequestResponseInformation.
ResponseTopic should be able to publish by all clients.

**Comment by redboltz** [comment]
> The client id should be unique on the broker.
> 
> So, we decided that client id within the broker should be: **username + user supplied client_id**

Yes. It violates MQTT spec but we choose this intentionally as we discussed.

> If the broker is generating the client id, it can generate a unique client id. The broker can generate a unique topic for request/reply based on unique client_id

Do you mean for auto client_id generation, uuid is reasonable choice but for request/response topic, there is a better choice ?

For request/response, `username + client_id` can be used. ( It not essential that we can add `response_` prefix )
client_id could be supplied by user, if it is empty, then broker generate it by uuid.
Anyway `username + client_id(user supplied or broker generated)` can be used as response topic.

Am I understanding correclty?

**Comment by redboltz** [comment]
I just worried about implementation complexity. But I think that your design doesn't need to big modification to support connection based authorization for request/response topic. It's nice!

Let's move on this way.

**Comment by kleunen** [comment]
But what you think of an additional method ('anonymous'), next to 'password' and 'client_cert' to specify anonymous user ? 

````
  "authentication": [
        {
            "name": "anonymous",
            "method": "anonymous"
        }
    ]

**Comment by redboltz** [comment]
We need to define what anonymous user/connection mean.

1. CONNECT packet UserName: "anonymous"
2. CONNECT without UserName fileds 
3. CONNECT with UserName: "" (empty string)

**Comment by redboltz** [comment]
Thanks. Do you treat 3 "" as the same as normal user like "u1" ?

**Comment by redboltz** [comment]
I think that now we are ready to implement.
Please let me know if I look over something.

**Comment by redboltz** [comment]
Additional information about Request/Response.
https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901252

> reply-topic-1 (connect attempt 1)
> reply-topic-2 (connect attempt 2)
> reply-topic-3 (connect attempt 3)

The behavior is up to broker's implementation but the following implementation is the most user friendly, I think.

When the broker receives CONNECT with RequestResponseInformation(1), then applies the following logic.

```cpp
if (clean_start) {
    allocate the new uuid for response_topic.
    store the topic to the session_state
}
else {
    if the session_state that is corresponding to the client_id (it will become UserName+ClientId), 
    and the session_state has response_topic, then return it.
    otherwise allocate the new uuid for response_topic and store it to session_state
}
```

So response_topic can exist while session_state exists. Similar to the normal topic, the response topic can be retained, storing message after disconnection, and restore the message on reconnection.

It is implemented by #894

**Comment by kleunen** [comment]
Yes, seems to make sense.
